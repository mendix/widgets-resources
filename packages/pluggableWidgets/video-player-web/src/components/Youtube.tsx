import { Component, createElement, createRef } from "react";
import ReactResizeDetector from "react-resize-detector";
import { fixHeightWithRatio, getRatio, validateUrl } from "../utils/Utils";

export interface YoutubeProps {
    url: string;
    autoPlay: boolean;
    showControls: boolean;
    loop: boolean;
    muted: boolean;
    aspectRatio?: boolean;
}

export interface YoutubeState {
    ratio: number;
}

export class Youtube extends Component<YoutubeProps, YoutubeState> {
    private iframe = createRef<HTMLIFrameElement>();
    private readonly handleOnResize = this.onResize.bind(this);
    private handleAttributes = this.getUrlAttributes.bind(this);
    readonly state: YoutubeState = {
        ratio: 0
    };

    render(): JSX.Element {
        return (
            <iframe
                className="widget-video-player-iframe"
                src={this.generateUrl(this.props.url)}
                frameBorder="0"
                allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
                allowFullScreen
                ref={this.iframe}
            >
                <ReactResizeDetector
                    handleWidth
                    handleHeight
                    onResize={this.handleOnResize}
                    refreshMode="debounce"
                    refreshRate={100}
                />
            </iframe>
        );
    }

    private onResize(): void {
        if (this.iframe && this.iframe.current && this.props.aspectRatio) {
            if (this.state.ratio) {
                fixHeightWithRatio(this.iframe.current, this.state.ratio);
            } else {
                getRatio(this.props.url).then(ratio => {
                    this.setState({ ratio });
                    if (this.iframe && this.iframe.current) {
                        fixHeightWithRatio(this.iframe.current, this.state.ratio);
                    }
                });
            }
        }
    }

    private generateUrl(url: string): string {
        const attributes = this.handleAttributes();
        try {
            if (url.includes("youtube.com/embed/")) {
                return `${url}${attributes}`;
            }
            if (url.includes("youtu.be/") || url.includes("youtube.com/v/")) {
                const urlSplit = url.split("/");
                if (urlSplit.length > 0) {
                    const id = urlSplit[urlSplit.length - 1];
                    return `https://www.youtube.com/embed/${id}${attributes}`;
                }
            }
            if (url.includes("youtube.com/watch?v=")) {
                const urlYoutubeSplit = url.split("watch?v=");
                if (urlYoutubeSplit.length > 0) {
                    const id = urlYoutubeSplit[urlYoutubeSplit.length - 1];
                    return `https://www.youtube.com/embed/${id}${attributes}`;
                }
            }
        } catch (e) {
            return url;
        }
        return url;
    }

    private getUrlAttributes(): string {
        let attributes = "?modestbranding=1&rel=0";
        attributes += "&autoplay=" + (this.props.autoPlay ? "1" : "0");
        attributes += "&controls=" + (this.props.showControls ? "1" : "0");
        attributes += "&muted=" + (this.props.muted ? "1" : "0");
        attributes += "&loop=" + (this.props.loop ? "1" : "0");
        return attributes;
    }

    // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
    public static canPlay(url: string): boolean {
        return !!url && !!validateUrl(url) && (url.indexOf("youtube.com") > -1 || url.indexOf("youtu.be") > -1);
    }
}
