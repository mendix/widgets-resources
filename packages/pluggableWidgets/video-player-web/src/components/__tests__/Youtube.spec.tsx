import { createElement } from "react";

import { Youtube, YoutubeProps } from "../Youtube";
import { create } from "react-test-renderer";
import ReactResizeDetector from "react-resize-detector";

describe("YoutubePlayer Player", () => {
    const defaultProps = {
        url: "http://youtube.com/watch?v=123456",
        autoPlay: false,
        muted: false,
        loop: false,
        showControls: false,
        aspectRatio: false
    };

    const defaulPlayer = (props: YoutubeProps): JSX.Element => <Youtube {...props} />;

    it("should renders correctly", () => {
        const player = create(defaulPlayer(defaultProps)).toJSON();

        expect(player).toMatchSnapshot();
    });

    it("should renders correctly with autoplay", () => {
        const player = create(defaulPlayer({ ...defaultProps, autoPlay: true })).toJSON();

        expect(player).toMatchSnapshot();
    });

    it("should renders correctly with muted", () => {
        const player = create(defaulPlayer({ ...defaultProps, muted: true })).toJSON();

        expect(player).toMatchSnapshot();
    });

    it("should renders correctly with controls", () => {
        const player = create(defaulPlayer({ ...defaultProps, showControls: true })).toJSON();

        expect(player).toMatchSnapshot();
    });

    it("should renders correctly with aspectRatio", () => {
        const player = create(defaulPlayer({ ...defaultProps, aspectRatio: true }));
        window.dispatchEvent(new Event("resize"));
        const instance = player.root;
        const sizeDetector = instance.findByType(ReactResizeDetector);

        expect(player.toJSON).toMatchSnapshot();
        expect(sizeDetector).not.toBeNull();
        expect(sizeDetector.props).toHaveProperty("onResize");
    });
});
