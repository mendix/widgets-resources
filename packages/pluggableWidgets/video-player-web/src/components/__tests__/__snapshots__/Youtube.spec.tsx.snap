// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YoutubePlayer Player should renders correctly 1`] = `
<iframe
  allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
  allowFullScreen={true}
  className="widget-video-player-iframe"
  frameBorder="0"
  src="https://www.youtube.com/embed/123456?modestbranding=1&rel=0&autoplay=0&controls=0&muted=0&loop=0"
>
  <div />
</iframe>
`;

exports[`YoutubePlayer Player should renders correctly with aspectRatio 1`] = `[Function]`;

exports[`YoutubePlayer Player should renders correctly with autoplay 1`] = `
<iframe
  allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
  allowFullScreen={true}
  className="widget-video-player-iframe"
  frameBorder="0"
  src="https://www.youtube.com/embed/123456?modestbranding=1&rel=0&autoplay=1&controls=0&muted=0&loop=0"
>
  <div />
</iframe>
`;

exports[`YoutubePlayer Player should renders correctly with controls 1`] = `
<iframe
  allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
  allowFullScreen={true}
  className="widget-video-player-iframe"
  frameBorder="0"
  src="https://www.youtube.com/embed/123456?modestbranding=1&rel=0&autoplay=0&controls=1&muted=0&loop=0"
>
  <div />
</iframe>
`;

exports[`YoutubePlayer Player should renders correctly with muted 1`] = `
<iframe
  allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
  allowFullScreen={true}
  className="widget-video-player-iframe"
  frameBorder="0"
  src="https://www.youtube.com/embed/123456?modestbranding=1&rel=0&autoplay=0&controls=0&muted=1&loop=0"
>
  <div />
</iframe>
`;
